### PR TITLE
Testing skeleton for pivot phase

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <chuckh@vmware.com>

You can skip all the code down to https://github.com/detiber/cluster-api/pull/1/files#diff-8b8b59673e2cc5a17cb627410440f363R216

it's the only interesting part. The rest is just mocking out a dumb client.

54% coverage with this one test! It can also be turned into table testing fairly easily depending on the conditions we want to test

### CODE REVIEW

File: cmd/clusterctl/phases/pivot.go [hunk 1]
```diff
@@ -25,28 +25,61 @@ import (
...
+	GetMachineSets(namespace string) ([]*clusterv1.MachineSet, error)
+	GetMachineSetsForCluster(string, string) ([]*clusterv1.MachineSet, error)
+	GetMachineSetsForMachineDeployment(*clusterv1.MachineDeployment) ([]*clusterv1.MachineSet, error)
+	GetMachinesForCluster(string, string) ([]*clusterv1.Machine, error)
+	GetMachinesForMachineSet(*clusterv1.MachineSet) ([]*clusterv1.Machine, error)
+	ScaleStatefulSet(string, string, int) error
+	WaitForClusterV1alpha1Ready() error
+}
+
+type targetClient interface {
```
Comment (detiber) [inline cmd/clusterctl/phases/pivot.go:51]: Would it make sense to push these interfaces up to clusterclient?

File: cmd/clusterctl/phases/pivot_test.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+	})
+	return s
+}
+
+// Associations are done via namespaces. If you create a machinedeployment in
+// the same namespace as a cluster then CAPI will assume it owns that machinedeployment
```
Comment (detiber) [inline cmd/clusterctl/phases/pivot_test.go]: The association via namespace doesn't hold true anymore (at least for the general codebase). I need to rebase to pick up the changes that include Vince's MachineSet/MachineDeployment ownerRef updates.

Association is now done both by Namespace and the same label association as is done for Machines

**Review by detiber — COMMENTED** [review]
This is awesome, I was way overthinking the testing. This is a much more straightforward approach than I was thinking of with trying to leverage the client-go Fake client :joy:

File: cmd/clusterctl/phases/pivot.go (hunk 1 above)
Comment (chuckha) [reply cmd/clusterctl/phases/pivot.go:51]: I don't have a use case to push this up. If you do then we can definitely consider it!

For now though, I'm using this as a local scope down to make testing and readability of this functionality a bit easier

**Comment by chuckha** [comment]
Got this up to date with your branch and added a test that shows an example of how to test an error case. We don't have to 100% code coverage but just the fact that we *can* when bugs crop up is nice to know.

**Pull request merged**